File: J9Ch/src/J_2_Binary_Search/Related_8/_63_Search_in_Rotated_Sorted_Array_II_Medium.cpp
/**
63
Search in Rotated Sorted Array II
*/

#include <vector>

using namespace std;

class Solution {
    /**
     * param A : an integer ratated sorted array and duplicates are allowed
     * param target :  an integer to be search
     * return : a boolean
     */
public:
    bool search(vector<int> &A, int target) {
        // write your code here
        int ans = -1;
        int len = A.size();
        int cnt = 0;
        for ( int i = 0; i < len; ++i) {
            if (A[i] == target) {
                ans = i;
            }
            if (i != 0)
                if (A[i] < A[i-1])
                    cnt ++;
        }
        if (cnt  > 1)
            ans = -ans;
        return (ans>=0);
    }
};

File: J9Ch/src/J_2_Binary_Search/Related_8/python_cpp/_63_Search_in_Rotated_Sorted_Array_II_Medium.py
'''/**
63
Search in Rotated Sorted Array II
'''

class Solution:
    """
    @param A : an integer ratated sorted array and duplicates are allowed
    @param target : an integer to be searched
    @return : a boolean
    """
    def search(self, A, target):
        # write your code here
        l, h = 0, len(A) - 1
        while (l <= h):
            m = l + ((h - l) >> 1)
            if A[m] == target:
                return True
            elif (A[m] > A[l] and target < A[m] and target >= A[l]) or (A[m] < A[l] and not (target <= A[h] and target > A[m])):
                h = m - 1
            else:
                l = m + 1
        return False